tools/xibuild/Main.cs

```diff
@@ -0,0 +1,220 @@
+				Console.WriteLine ();
+				Console.WriteLine ("\t-c: Generate config file only");
+				Console.WriteLine ("\t-t: Path to the managed tool to run. If this and `-c` are not used, then this runs msbuild");
+				Console.WriteLine ("\t-m <base config file>: Config file to merge with the one from msbuild.dll.config");
+				Console.WriteLine ("\t-h: help");
+				Console.WriteLine ("\nNote: Adds the path from the environment variable MSBuildExtensionsPathFallbackPathsOverride to the list of fallback paths in the generated app.config");
+				Console.WriteLine ("\nExamples:");
+				Console.WriteLine ("xibuild -c /path/to/foo.exe : Generate /path/to/foo.exe.config");
+				Console.WriteLine ("xibuild -- /v:diag /path/to/project.csproj : Run msbuild with a custom app.config and the arguments passed to msbuild.");
+				Console.WriteLine ("xibuild -t -- /path/to/managed_tool.exe args : Run managed_tool.exe with the arguments, and a custom app.config");
+				Console.WriteLine ("");
+				Console.WriteLine ("Add `-m /path/to/base.exe.config` to merge the generated app.config with base.exe.config .");
+			}
+		}
+
+		static int RunTool (string toolPath, string combinedArgs, string baseConfigFile)
+		{
+			var tmpMSBuildExePathForConfig = Path.GetTempFileName ();
+			var configFilePath = tmpMSBuildExePathForConfig + ".config";
+
+			GenerateAppConfig (configFilePath, baseConfigFile);
+
+			// Required so that msbuild can read the correct config file
+			Environment.SetEnvironmentVariable ("MSBUILD_EXE_PATH", tmpMSBuildExePathForConfig, EnvironmentVariableTarget.Process);
+
+			var p = Process.Start (new ProcessStartInfo {
+				FileName = toolPath,
+				Arguments = combinedArgs,
+				UseShellExecute = false,
+				RedirectStandardInput = true,
+			});
+
+			p.WaitForExit ();
+			return p.ExitCode;
+		}
+
+		static void GenerateAppConfig (string targetConfigFile, string baseConfigFile)
+		{
+			bool IsMacOS = System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform (System.Runtime.InteropServices.OSPlatform.OSX);
+
+			//FIXME: Current? or from PATH?
+			string mono = IsMacOS ? "/Library/Frameworks/Mono.framework/Versions/Current/lib/mono" : "/usr/lib/mono";
+			string vsVersion = "15.0";
+			string MSBuildBin = Path.Combine (mono, "msbuild", vsVersion, "bin");
+
+			string MSBuildConfig = Path.Combine (MSBuildBin, "MSBuild.dll.config");
+			string MSBuildExtensionsPath = Path.Combine (mono, "xbuild");
+			string FrameworksDirectory = Path.Combine (mono, "xbuild-frameworks");
+			string MSBuildSdksPath = Path.Combine (MSBuildBin, "Sdks");
+
+			var dstXml = new XmlDocument ();
+
+			var dstConfigNode = dstXml.CreateNode (XmlNodeType.Element, "configuration", "");
+			dstXml.AppendChild (dstConfigNode);
+
+			if (!String.IsNullOrEmpty (baseConfigFile) && File.Exists (baseConfigFile)) {
+				var baseXml = new XmlDocument ();
+				baseXml.Load (baseConfigFile);
+				CopyConfigNode (baseXml, dstConfigNode);
+			}
+
+			// Copy over msbuild.dll.config
+			{
```

Comment:
Why the {} ?

Reply:
There are multiple *Xml variables, and this one is required only for this specific part, so scoped it to prevent accidental use.